feature/schedule_entity

HyunJng opened the pull request:
- Account, Daywork, Schedule 엔티티 객체 생성
- AccountType, DayworkPriority, Workcheck Enume객체 생성

나중에 Entity 바뀔 것 같지만 일단 제가 설계한대로는 DB도 제대로 생성되는 것 확인까지 했습니다.

HyunJng committed:
feat([Entity related Schedule]): Schedule, Account, Daywork 엔티티 설계

- Account, Daywork, Schedule 엔티티 객체 생성
- AccountType, DayworkPriority,
Workcheck Enume객체 생성

JinhyeonKwak commented on src/main/java/com/whatachad/app/model/domain/Schedule.java:
```diff
@@ -0,0 +1,36 @@
+import java.util.List;
+
+@Builder
+@NoArgsConstructor
```
저는 생성자를 열 때 protected를 쓰는 편입니다. 생성자를 public하게 열어 놓으면 안전한 Builder 패턴을 사용하는 이유가 퇴색될 우려가 있습니다. 협업하는 개발자가 생성자를 이용하여 엔티티를 생성하는 과정에서 실수로 파라미터 값의 순서를 바꿔서 넣는 등의 문제가 발생하면 원치않는 오류가 발생할 수도 있을 것 같아요. 그래서 저는 protected를 써서 같은 패키지의 test 코드까지는 쓸 수 있도록 허용하게 만들고 있어요.

JinhyeonKwak commented on src/main/java/com/whatachad/app/model/domain/Schedule.java:
```diff
@@ -0,0 +1,36 @@
+
+@Builder
+@NoArgsConstructor
+@AllArgsConstructor
```
이 친구도 마찬가지입니다 ㅎㅎ

JinhyeonKwak commented on src/main/java/com/whatachad/app/model/domain/Account.java:
```diff
@@ -0,0 +1,32 @@
+    @Enumerated(EnumType.STRING)
+    private AccountType type;
+
+    private int cost;
```
primitive type인 int가 아닌 Integer 타입을 쓰는 것은 어떻게 생각하시나요? Integer를 쓰게 되면 null을 허용함으로써 실수로 cost를 저장하지 못하였을 때에도 예외를 터뜨릴 수 있지만 int를 쓰면 cost=0인 값이 저장되므로 의도치 않은 결과를 보여줄 수 있을 것 같아요!
뿐만 아니라 Integer가 구현하고 있는 여러 메서드를 이용할 수 있으므로 형변환 없이 바로바로 사용할 수 있다는 장점이 있을 수 있겠네요!

JinhyeonKwak commented on src/main/java/com/whatachad/app/model/domain/Account.java:
```diff
@@ -0,0 +1,32 @@
+
+    private int cost;
+
+    private int time;
```
의도하신 time 필드가 수입/지출 내역의 시간을 의미하는 거라면, 08:24처럼 시간을 표현을 못하게 되겠군요.
어쨌든 시간을 표현해야 한다면 몇 가지 방법이 있을텐데, 바로 떠오르는 것은 LocalDateTime이나 Integer/String 등으로 직접 표현하는 것이겠죠. 
하지만 LocalDateTime을 쓰는 경우에 이 값을 json의 string 형태로 내려준다고 가정한다면, 클라이언트 측에서는 이를 적절히 parsing해서 써야 할 거에요. 왜냐하면 사용자에게 23:52:35, 2023-04-08T12:32:22.000003333 등의 형태로 보여주고 싶지는 않을거니깐 말이죠. 그래서 서버에서 parsing을 한 다음에 데이터를 건네주어야 할 것이고, 그렇다면 미리 가공된 데이터를 필드로서 가지고 있으면 더 편리하게 사용할 수 있겠다는 생각이 듭니다.
그래서 저는 DateTime이라는 값 타입을 따로 만들어서 내장 타입으로 사용할 것 같습니다.
```
@Embeddable
public class DateTime {
    private Integer date;
    private Integer hour;
    private Integer minute;
}

public class Account {
    ~
    @Embedded
    private DateTime dateTime;
    ~
}
```

JinhyeonKwak commented on src/main/java/com/whatachad/app/model/domain/Daywork.java:
```diff
@@ -0,0 +1,34 @@
+    private Workcheck status;
+
+    int time;
+    int date;
```
여기서도 내장 타입을 재활용할 수 있겠네요 ㅎㅎ

JinhyeonKwak commented on src/main/java/com/whatachad/app/model/domain/Workcheck.java:
```diff
@@ -0,0 +1,5 @@
+package com.whatachad.app.model.domain;
+
+public enum Workcheck {
+    COMPLETE, YET
```
YET -> NOT_YET or NOT_COMPLETE 도 한 번 고려해 주세요.

JinhyeonKwak replied:
그리고 저는 지금 enum class들을 type 패키지에서 한 번에 관리하고 있는데, 이 부분은 어떻게 생각하시나요?

HyunJng committed:
refactor([Entity related Schedule]): 코드리뷰 수정사항 반영

1. Account, Daywork, Schedule 엔티티
	- constructor의 access레벨 : public ->protected
	- primitive type -> Wrapper type

2. Daywork, Schedule엔티티
	- 날짜/시간관련 필드를 DateTime클래스 생성하여 embedded

3. schedule 엔티티
	- List타입의 필드에 cascade 속성 추가

3. AccountType, DayworkPriority, Workcheck enum클래스
	- 이동 : domain패키지 -> type패키지
